README.md
django-flatpages-i18n
=====================

Translatable version of django.contrib.flatpages with basic menu support.

Installation
-------------

```
$ pip install django-flatpages-i18n
```

Add new applications at the end of INSTALLED_APPS in your settings.py.

```python
INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.admin',
    'django.contrib.admindocs',

    'south',

    'mptt',
    'modeltranslation',
    'flatpages_i18n',
)
```

Before the migration we need create the list of all available languages in settings.py.

```python
LANGUAGE_CODE = 'en'

from django.utils.translation import gettext

LANGUAGES = (
    ('en', gettext('English')),
    ('de', gettext('German')),
)
```

Add the FlatpageFallbackMiddleware at the end of MIDDLEWARE_CLASSES.

```python
MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.middleware.locale.LocaleMiddleware', # Don't forget
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',

    'flatpages_i18n.middleware.FlatpageFallbackMiddleware' # Add this
)
```

Run the migrations.

```
$ python manage.py migrate flatpages_i18n
```


Usage
-----

###  To get all flatpages:

```python
{% load i18n flatpages_i18n %}

{% get_available_languages as LANGUAGES %}
{% get_flatpages_i18n as flatpages_i18n %}

<ul>
    {% for flatpage in flatpages_i18n %}
        <li><a href="/{{ LANGUAGE_CODE }}{{ flatpage.url }}">{{ flatpage.title }}</a></li>
    {% endfor %}
</ul>
```

To get flatpage by its PK:

```python
{% get_flatpage_i18n 123 as my_flatpage %}
{{ my_flatpage.content }}
```

or by its machine_name:

```python
{% get_flatpage_i18n 'my-flatpage' as my_flatpage %}
```

### Menu system:

```python
<div id="navigation">
    {% get_menu %}
</div>
```

get only children of menu item identified by its PK

```python
<div id="navigation">
    {% get_menu 2 %}
</div>
```

or by its machine_name

```python
<div id="navigation">
    {% get_menu 'footer-menu' %}
</div>
```


HISTORY.rst
History
=========

0.1.3 (2013-10-11)

* flatpages are able to put into trees by using ``django-mptt`` application
* basic menu system support
* removed comment field for flatpages

0.2.0 (2014-01-17)
* refactored flatpages > flatpages_i18n, update your template tags
* menu can be called also by its machine_name

0.2.1 (2014-01-17)
* uniqueness of MenuItem's machine_name

0.2.2 (2014-02-28)
* Django 1.6 support

0.3.0 (2014-03-04)
* get_flatpage_i18n template tag, use migration.

0.4.0 (2014-05-20)
* Django 1.8 support

0.4.4 (2017-03-01)
* Python 3 support

0.5.0 (2017-03-02)
* translation fields are synced automatically after initial migration

0.6.0 (2017-05-15)
* target of menu items

0.7.0 (2017-05-22)
* language code added to absolute path of FlatPage_i18n model
* support for django-seo2 app

0.7.1 (2018-01-24)
* support for Django 1.11

0.7.2
* support for languages with hyphens

1.0.0 (2020-06-19)
* Complete refactoring of whole package.
* Dropped support for Python 2
* django-modeltrans replaced django-modeltranslations
* flatpage doesn't have parent relationship
* dropped middleware, using urls instead
* dropped machine name for menu items
* dropped support for redactor editor
* GIN indexes
* WARNING: migrations are not compatible with versions < 1.0.0 !

flatpages_i18n/migrations/0001_initial.py
# Generated by Django 2.2.4 on 2020-06-19 08:57

import django.contrib.postgres.indexes
from django.db import migrations, models
import django.db.models.deletion
import flatpages_i18n.models
import modeltrans.fields
import mptt.fields


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('sites', '0002_alter_domain_unique'),
    ]

    operations = [
        migrations.CreateModel(
            name='FlatPage_i18n',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('title', models.CharField(max_length=200, verbose_name='title')),
                ('slug', models.SlugField(blank=True, default='', max_length=150, unique=True)),
                ('machine_name', models.SlugField(blank=True, help_text='unique', max_length=30, verbose_name='machine name')),
                ('content', models.TextField(blank=True, verbose_name='content')),
                ('template_name', models.CharField(blank=True, help_text="Example: 'flatpages_i18n/contact_page.html'. If this isn't provided, the system will use 'flatpages_i18n/default.html'.", max_length=70, verbose_name='template name')),
                ('registration_required', models.BooleanField(default=False, help_text='If this is checked, only logged-in users will be able to view the page.', verbose_name='registration required')),
                ('created', models.DateTimeField(auto_now_add=True, verbose_name='created')),
                ('modified', models.DateTimeField(auto_now=True, verbose_name='modified')),
                ('i18n', modeltrans.fields.TranslationField(fields=('title', 'slug', 'content'), required_languages=(), virtual_fields=True)),
                ('sites', models.ManyToManyField(to='sites.Site')),
            ],
            options={
                'verbose_name': 'flat page',
                'verbose_name_plural': 'flat pages',
                'ordering': ('title',),
            },
            bases=(flatpages_i18n.models.SlugMixin, models.Model),
        ),
        migrations.CreateModel(
            name='Menu',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('machine_name', models.SlugField(unique=True)),
            ],
            options={
                'verbose_name': 'menu',
                'verbose_name_plural': 'menus',
            },
        ),
        migrations.CreateModel(
            name='MenuItem',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('custom_link', models.CharField(blank=True, max_length=200, verbose_name='custom link')),
                ('target', models.CharField(choices=[('_blank', 'Opens the linked document in a new window or tab'), ('_self', 'Opens the linked document in the same frame as it was clicked'), ('_parent', 'Opens the linked document in the parent frame'), ('_top', 'Opens the linked document in the full body of the window')], default='_self', max_length=7, verbose_name='target')),
                ('title', models.CharField(blank=True, max_length=255, verbose_name='title')),
                ('created', models.DateTimeField(auto_now_add=True, verbose_name='created')),
                ('modified', models.DateTimeField(auto_now=True, verbose_name='modified')),
                ('i18n', modeltrans.fields.TranslationField(fields=('title', 'custom_link'), required_languages=(), virtual_fields=True)),
                ('lft', models.PositiveIntegerField(editable=False)),
                ('rght', models.PositiveIntegerField(editable=False)),
                ('tree_id', models.PositiveIntegerField(db_index=True, editable=False)),
                ('level', models.PositiveIntegerField(editable=False)),
                ('flatpage', models.ForeignKey(blank=True, default=None, null=True, on_delete=django.db.models.deletion.SET_NULL, to='flatpages_i18n.FlatPage_i18n', verbose_name='flatpage')),
                ('menu', models.ForeignKey(on_delete=django.db.models.deletion.PROTECT, related_name='item_set', to='flatpages_i18n.Menu', verbose_name='menu')),
                ('parent', mptt.fields.TreeForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='children', to='flatpages_i18n.MenuItem')),
            ],
            options={
                'verbose_name': 'menu item',
                'verbose_name_plural': 'menu items',
            },
        ),
        migrations.AddIndex(
            model_name='menuitem',
            index=django.contrib.postgres.indexes.GinIndex(fields=['i18n'], name='flatpages_i_i18n_4204f4_gin'),
        ),
        migrations.AddIndex(
            model_name='flatpage_i18n',
            index=django.contrib.postgres.indexes.GinIndex(fields=['i18n'], name='flatpages_i_i18n_ea673c_gin'),
        ),
    ]
